### ISSUE TITLE
Update dev/Dockerfile to use gcc 4.8 by default

### ISSUE THREAD

**Issue by henrytansetiawan** [tensorflow/io#173]
The default gcc installed on the Docker image is gcc 7.0.3. Compiling custom ops using newer gcc and interop-ing with older gcc 4.8 used by official TF build prones to Segmentation Fault ABI issue: https://github.com/tensorflow/tensorflow/issues/13308.

some reference:
- https://gist.github.com/application2000/73fd6f4bf1be6600a2cf9f56315a2d91
- https://gist.github.com/vmaks/06ac9bcd859621aa589f

**Comment by yongtang** [comment]
@henrytansetiawan I think this is because the tfio-dev uses Ubuntu 18.04 as the base image. In our Travis CI build, we use Ubuntu 14.04 which matches with tensorflow's build. Changing the tfio-dev's base image to Ubuntu 14.04 should fix the issue I believe.

**Comment by henrytansetiawan** [comment]
@yongtang sounds good, that would do and a simpler fix.

**Comment by yongtang** [comment]
@henrytansetiawan PR #189 is another victim of the ABI compatibility issue.

Also /cc @BryanCutler for the Dockerfile as well.

We should always stick with Ubuntu 14.04. However, there are some underlying obstacles:
- Tensorflow's binary is built with Ubuntu 14.04 for 2.7-3.6, but seems to be Ubuntu 16.04 for 3.7.
- Tensorflow's binary is built with system default python 2.7 and 3.4 for Ubuntu 14.04.
- Tensorflow's binary is built build with deprecated `ppa:fkrull/deadsnakes`(not the current `deadsnakes` ppa)  for Python 3.5 on Ubuntu 14.04. See [link](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/install/install_python3.5_pip_packages.sh) here.
- Tensorflow's binary is built build with compiled python of source code for Python 3.6.1 on Ubuntu 14.04. See [link](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/tools/ci_build/install/install_python3.6_pip_packages.sh) here.
- Ubuntu 14.04's system pip (`apt install python-pip`) does not work with tensorflow, `python get-pip.py` has to be used.
- Ubuntu 14.04 does not have python 3.7 and deadsnake 's version does not work on Ubuntu 14.04 either. This is blocking us from releasing 3.7 version of tensorflow-io.
- To build the wheel file we also need `patchelf` which does not comes with Ubuntu 14.04, so we build patchelf from source for Travis CI.
- `auditwheel` is also needed to build wheel. However, `auditwheel` is only available in Python 3. In Travis CI, we have a very awkward workaround of build binary for python 2, then patch and use python 3 of auditwheel.

At the moment we just build one version of the binary `.so` file, and distributed with different python versions. In theory, we only need the header files from tensorflow library, and a gcc to build the binary. The python dependency is not needed.

However, we could only get header files from the released whl file in tensorflow, which requires python and pip install. It complicates things as python and pip installs some (unintended) libraries in the system.

I had some success to unzip the whl files and extract the header files. Still some remaining work to do but close.

To summerize:
- I think we should provide a Dockerfile with Ubuntu 14.04 (or the version chosen by tensorflow build), and then package needed python install.
- We should also provided a way to build tensorflow-io with tensorflow source in the same environment, so that we could avoid issues like #161 and #188.


Also, in our CI, we have test setup for Ubuntu 16.04 and 18.04 which have different gcc versions (than tensorflow bianry). I think we may need to drop this idea. We could setup tensorflow+tensorflow-io source build instead (if we could get Google CI for heavy build support).

**Comment by yongtang** [comment]
Another issue is that tf 1.14 will be released in several days (expected), but 1.14 may have changed the `.so` to `.so.1` (See PR #190 which fails with tf-nightly). We will need to see what is the format released by 1.14.

Also /cc @terrytangyuan

**Comment by BryanCutler** [comment]
Yeah, I agree it would probably be best to use 14.04 for the dev image

**Comment by yongtang** [comment]
Adding https://github.com/tensorflow/tensorflow/issues/27067 as a reference here as well. Once C API is in place, we may need to rework on some of the ops I guess.

**Comment by BryanCutler** [comment]
@yongtang what do you think about using `tensorflow/tensorflow:custom-op` as the base for a dev image?  I tried it out and seems to work well so I can submit the change.

**Comment by yongtang** [comment]
@BryanCutler That would be great! The `tensorflow/tensorflow:custom-op` has the same environment as tensorflow's build env so binary wise it works. Though some packages may have been outdated. For example, `bazel` version in custom-op is still 0.15.0 which will need to be updated to 0.24.1 for tensorflow to build. But otherwise it works ok.

**Comment by zh794390558** [comment]
@yongtang  So this fix is for tf 1.14 ?
